node/overseer/src/lib.rs

```diff
@@ -343,33 +343,56 @@ impl<M: Send + 'static> SubsystemContext for OverseerSubsystemContext<M> {
 		}).await.map_err(Into::into)
 	}
 
-	async fn send_message(&mut self, msg: AllMessages) {
+	async fn send_message(&mut self, msg: AllMessages) -> bool {
 		self.send_and_log_error(ToOverseer::SubsystemMessage(msg)).await
 	}
 
-	async fn send_messages<T>(&mut self, msgs: T)
+	async fn send_messages<T>(&mut self, msgs: T) -> bool
 		where T: IntoIterator<Item = AllMessages> + Send, T::IntoIter: Send
 	{
-		let mut msgs = stream::iter(msgs.into_iter().map(ToOverseer::SubsystemMessage).map(Ok));
-		if self.tx.send_all(&mut msgs).await.is_err() {
-			tracing::debug!(
-				target: LOG_TARGET,
-				msg_type = std::any::type_name::<M>(),
-				"Failed to send messages to Overseer",
-			);
+		const SEND_TIMEOUT: Duration = Duration::from_millis(1000);
 
+		let mut msgs = stream::iter(msgs.into_iter().map(ToOverseer::SubsystemMessage).map(Ok));
+		match self.tx.send_all(&mut msgs).timeout(SEND_TIMEOUT).await {
+			None => {
+				tracing::error!(target: LOG_TARGET, "Outgoing messages timeout reached");
+				false
+			}
+			Some(Err(_)) => {
+				tracing::debug!(
+					target: LOG_TARGET,
+					msg_type = std::any::type_name::<M>(),
+					"Failed to send a messages to Overseer",
+				);
+
+				// We return `true` here because errors indicate no change of re-try.
```

Comment:
we could use `try_send` instead of timeouts, this way, we could propagate the messages in case of the buffer being full, otherwise the caller must always clone the messages if they want to retry

Reply:
I see what you're getting at, but I think that's a much larger refactoring than I wanted to commit to at the moment. If we wait for a small amount of time then the odds that buffers drain during that time is much higher. Whereas `try_send` is likely to encounter failure much more often and would imply rewriting a lot of subsystem logic to do retrying.